In [10]:
import pandas as pd, numpy as np

RAW   = "sat-results.csv"
CLEAN = "cleaned_sat_results.csv"

# 1) Read robustly (auto-detect delimiter ; or , and skip bad lines)
df_raw = pd.read_csv(RAW, sep=None, engine="python", on_bad_lines="skip")

# 2) Normalize headers and pick required columns
def norm(s):
    return str(s).strip().lower().replace(".", "").replace("%","").replace("-", " ").replace("/", " ").replace("_"," ").replace("  "," ")

norm_cols = {c: norm(c) for c in df_raw.columns}

wanted = {
    "dbn": ["dbn"],
    "school_name": ["school name"],
    "num_of_test_takers": ["num of sat test takers","number of sat test takers","num sat test takers"],
    "pct_students_tested": ["pct students tested","percent students tested"],
    "critical_reading_avg": ["sat critical reading avg score","sat critical readng avg score"],
    "math_avg": ["sat math avg score"],
    "writing_avg": ["sat writing avg score"],
}

pick = {}
for std, variants in wanted.items():
    found = None
    for actual, n in norm_cols.items():
        if n in variants:
            found = actual
            break
    if not found:
        raise SystemExit(f"Missing column for: {std}\nAvailable: {list(df_raw.columns)}")
    pick[std] = found

df = df_raw[[pick[k] for k in wanted.keys()]].rename(columns={v:k for k,v in pick.items()})

# 3) Numeric cleaning
def to_number(x):
    if pd.isna(x): return np.nan
    s = str(x).strip().replace(",", "").replace("%","")
    if s.lower() in ("", "na", "n/a", "null", "none"): return np.nan
    try: return float(s)
    except: return np.nan

for c in ["num_of_test_takers","critical_reading_avg","math_avg","writing_avg"]:
    df[c] = df[c].map(to_number)

df["pct_students_tested"] = pd.to_numeric(
    df["pct_students_tested"].astype(str).str.replace("%","", regex=False).str.strip(),
    errors="coerce"
)

# Valid SAT range
for c in ["critical_reading_avg","math_avg","writing_avg"]:
    df.loc[~df[c].between(200, 800), c] = np.nan

df = df.drop_duplicates(subset=["dbn"])

# 4) Save cleaned CSV and preview
df.to_csv(CLEAN, index=False)
print("Saved:", CLEAN, "| Rows:", len(df), "| Cols:", list(df.columns))
df.head()


Saved: cleaned_sat_results.csv | Rows: 478 | Cols: ['dbn', 'school_name', 'num_of_test_takers', 'pct_students_tested', 'critical_reading_avg', 'math_avg', 'writing_avg']


dbn                                    school_name  num_of_test_takers  \
0  01M292  HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES                29.0   
1  01M448            UNIVERSITY NEIGHBORHOOD HIGH SCHOOL                91.0   
2  01M450                     EAST SIDE COMMUNITY SCHOOL                70.0   
3  01M458                      FORSYTH SATELLITE ACADEMY                 7.0   
4  01M509                        MARTA VALLE HIGH SCHOOL                44.0   

   pct_students_tested  critical_reading_avg  math_avg  writing_avg  
0                 78.0                 355.0     404.0        363.0  
1                  NaN                 383.0     423.0        366.0  
2                  NaN                 377.0     402.0        370.0  
3                 92.0                 414.0     401.0        359.0  
4                 92.0                 390.0     433.0        384.0

In [11]:
'postgresql://neondb_owner:npg_7n5epBJCYdiG@ep-calm-paper-agvqn80r-pooler.c-2.eu-central-1.aws.neon.tech/neondb?sslmode=require&channel_binding=require'


'postgresql://neondb_owner:npg_7n5epBJCYdiG@ep-calm-paper-agvqn80r-pooler.c-2.eu-central-1.aws.neon.tech/neondb?sslmode=require&channel_binding=require'

In [13]:
!pip -q install pandas SQLAlchemy psycopg2-binary

import pandas as pd, re
from sqlalchemy import create_engine, text


neon_database_url = "postgresql://neondb_owner:npg_7n5epBJCYdiG@ep-calm-paper-agvqn80r-pooler.c-2.eu-central-1.aws.neon.tech/neondb?sslmode=require&channel_binding=require"

sqlalchemy_url = re.sub(r"^postgresql://", "postgresql+psycopg2://", neon_database_url.strip())
engine = create_engine(sqlalchemy_url, pool_pre_ping=True)

df = pd.read_csv("cleaned_sat_results.csv")
table_name = "sat_results_robert"
df.to_sql(table_name, engine, if_exists="replace", index=False)

with engine.connect() as conn:
    rows = conn.execute(text(f"SELECT COUNT(*) FROM {table_name}")).scalar()
    print("Rows in table:", rows)

engine.dispose()


Rows in table: 478


In [21]:
from sqlalchemy import text

with engine.begin() as conn:
    conn.execute(text("""
        DROP VIEW IF EXISTS sat_kpis_robert CASCADE;
        DROP VIEW IF EXISTS sat_enriched_robert CASCADE;
        DROP TABLE IF EXISTS sat_model_robert CASCADE;

        CREATE TABLE sat_model_robert AS
        SELECT
            dbn,
            school_name,
            CAST(num_of_test_takers AS INT)        AS num_of_test_takers,
            CAST(pct_students_tested AS FLOAT)     AS pct_students_tested,
            CAST(critical_reading_avg AS INT)      AS critical_reading_avg,
            CAST(math_avg AS INT)                  AS math_avg,
            CAST(writing_avg AS INT)               AS writing_avg
        FROM sat_results_robert
        WHERE num_of_test_takers IS NOT NULL;
    """))
print("sat_model_robert recreated successfully.")


sat_model_robert recreated successfully.


In [18]:
from sqlalchemy import text
import pandas as pd

sql_index = """
CREATE INDEX IF NOT EXISTS idx_sat_model_robert_dbn
ON sat_model_robert(dbn);
"""

sql_view = """
CREATE OR REPLACE VIEW sat_enriched_robert AS
WITH base AS (
    SELECT
        dbn,
        school_name,
        num_of_test_takers,
        pct_students_tested,
        critical_reading_avg,
        math_avg,
        writing_avg,
        ROUND( (critical_reading_avg + math_avg + writing_avg) / 3.0, 1) AS total_avg
    FROM sat_model_robert
),
ranked AS (
    SELECT
        b.*,
        PERCENT_RANK() OVER (ORDER BY total_avg) AS pr_total
    FROM base b
)
SELECT
    *,
    CASE
        WHEN pr_total >= 0.90 THEN 'A'
        WHEN pr_total >= 0.75 THEN 'B'
        WHEN pr_total >= 0.50 THEN 'C'
        WHEN pr_total >= 0.25 THEN 'D'
        ELSE 'E'
    END AS performance_band,
    CASE
        WHEN pct_students_tested IS NULL THEN 'Unknown'
        WHEN pct_students_tested >= 80 THEN 'High'
        WHEN pct_students_tested >= 50 THEN 'Medium'
        ELSE 'Low'
    END AS participation_band
FROM ranked;
"""

with engine.begin() as conn:
    conn.execute(text(sql_index))
    conn.execute(text(sql_view))

# quick checks
pd.read_sql("SELECT COUNT(*) AS rows FROM sat_enriched_robert;", engine)


rows
0   478

In [17]:
pd.read_sql("SELECT * FROM sat_enriched_robert ORDER BY total_avg DESC LIMIT 5;", engine)


dbn                                        school_name  \
0  29Q326                            CAMBRIA HEIGHTS ACADEMY   
1  16K688             THE BROOKLYN ACADEMY OF GLOBAL FINANCE   
2  02M473                             WASHINGTON IRVING YABC   
3  29Q283  PREPARATORY ACADEMY FOR WRITERS: A COLLEGE BOA...   
4  10X319  PROVIDING URBAN LEARNERS SUCCESS IN EDUCATION ...   

   num_of_test_takers  pct_students_tested  critical_reading_avg math_avg  \
0                   0                 78.0                   NaN     None   
1                   0                 85.0                   NaN     None   
2                   0                 92.0                   NaN     None   
3                  43                 85.0                 370.0     None   
4                   0                 92.0                   NaN     None   

   writing_avg total_avg  pr_total performance_band participation_band  
0          NaN      None  0.872117                B             Medium  
1          NaN      None  0.872117                B               High  
2          NaN      None  0.872117                B               High  
3        363.0      None  0.872117                B               High  
4          NaN      None  0.872117                B               High

In [25]:
from sqlalchemy import text

sql_view = """
CREATE OR REPLACE VIEW sat_enriched_robert AS
WITH base AS (
  SELECT
    dbn,
    school_name,
    num_of_test_takers,
    pct_students_tested,
    critical_reading_avg,
    math_avg,
    writing_avg,
    ROUND( (COALESCE(critical_reading_avg,0)
          + COALESCE(math_avg,0)
          + COALESCE(writing_avg,0)) / 3.0, 1) AS total_avg
  FROM sat_model_robert
),
ranked AS (
  SELECT
    b.*,
    PERCENT_RANK() OVER (ORDER BY total_avg) AS pr_total
  FROM base b
)
SELECT
  *,
  CASE
    WHEN pr_total >= 0.90 THEN 'A'
    WHEN pr_total >= 0.75 THEN 'B'
    WHEN pr_total >= 0.50 THEN 'C'
    WHEN pr_total >= 0.25 THEN 'D'
    ELSE 'E'
  END AS performance_band,
  CASE
    WHEN pct_students_tested IS NULL THEN 'Unknown'
    WHEN pct_students_tested >= 80 THEN 'High'
    WHEN pct_students_tested >= 50 THEN 'Medium'
    ELSE 'Low'
  END AS participation_band
FROM ranked;
"""
with engine.begin() as conn:
    conn.execute(text(sql_view))


In [32]:
import pandas as pd
from sqlalchemy import text

q_boro = """
SELECT
  LEFT(dbn, 2)                 AS borough_code,
  COUNT(*)                     AS total_schools,
  ROUND(AVG(total_avg), 1)     AS avg_total,
  ROUND(AVG(num_of_test_takers), 1) AS avg_takers,
  ROUND(AVG(critical_reading_avg), 1) AS avg_read,
  ROUND(AVG(math_avg), 1)         AS avg_math,
  ROUND(AVG(writing_avg), 1)      AS avg_write
FROM sat_enriched_robert
GROUP BY 1
ORDER BY 1;
"""
borough_summary = pd.read_sql(q_boro, engine)
borough_summary.to_csv("borough_performance.csv", index=False)
borough_summary


borough_code  total_schools  avg_total  avg_takers  avg_read  avg_math  \
0            01              9      435.3        73.3     423.8     468.4   
1            02             58      428.3       106.5     424.1     441.3   
2            03             13      413.2       108.8     424.8     430.8   
3            04              7      397.0        99.4     393.1     405.3   
4            05              9      406.1        70.4     407.3     409.7   
5            06             10      384.0        73.3     381.0     394.3   
6            07             15      377.4        56.3     378.1     381.7   
7            08             15      380.2        66.9     382.5     381.5   
8            09             21      375.0        55.2     372.5     379.6   
9            10             26      393.6       105.5     395.4     410.4   
10           11             17      386.0        74.5     387.5     391.8   
11           12             15      362.6        39.4     362.1     370.1   
12           13             16      409.7       137.0     411.9     414.5   
13           14             14      390.8        78.2     394.7     393.7   
14           15             11      376.9        58.5     391.5     392.3   
15           16              4      362.5        58.3     363.0     367.3   
16           17             18      382.3        69.6     384.9     388.4   
17           18             14      374.4        39.2     378.7     374.2   
18           19             12      367.9        63.2     367.1     377.6   
19           20              5      418.4       358.2     393.0     477.2   
20           21             11      400.0       191.0     395.9     417.2   
21           22              5      463.1       393.6     455.4     480.6   
22           23              6      361.7        36.5     358.7     375.2   
23           24             13      414.0       143.4     405.8     434.0   
24           25             11      433.2       140.0     423.0     458.5   
25           26              5      459.2       607.8     445.2     487.6   
26           27             11      405.4       145.7     405.6     419.8   
27           28             11      434.4       199.7     442.5     472.7   
28           29              9      376.2        54.4     391.8     400.0   
29           30              8      446.4       193.3     434.0     471.8   
30           31             12      453.1       252.1     453.8     462.9   
31           32              7      369.1        52.3     369.7     376.0   
32           75              1      435.3        32.0     429.0     444.0   
33           79              2      403.2         8.5     431.5     385.0   

    avg_write  
0       413.6  
1       419.4  
2       417.1  
3       392.7  
4       401.2  
5       376.7  
6       372.5  
7       376.6  
8       372.8  
9       390.8  
10      378.6  
11      355.5  
12      402.6  
13      384.0  
14      382.4  
15      357.3  
16      373.6  
17      370.4  
18      359.2  
19      385.0  
20      386.8  
21      453.4  
22      351.3  
23      402.2  
24      418.1  
25      444.8  
26      390.6  
27      431.0  
28      381.1  
29      433.5  
30      442.4  
31      361.6  
32      433.0  
33      393.0

In [26]:
import pandas as pd
pd.read_sql("""
  SELECT performance_band, COUNT(*) AS n
  FROM sat_enriched_robert
  GROUP BY performance_band
  ORDER BY performance_band;
""", engine)


performance_band    n
0                A   43
1                B   63
2                C  102
3                D  108
4                E  105

In [27]:
# Q6. Are there correlations between SAT subject scores?
q6 = """
SELECT
  ROUND(CORR(critical_reading_avg, math_avg)::numeric, 3)    AS corr_read_math,
  ROUND(CORR(critical_reading_avg, writing_avg)::numeric, 3) AS corr_read_write,
  ROUND(CORR(math_avg, writing_avg)::numeric, 3)             AS corr_math_write
FROM sat_enriched_robert;
"""
print("Q6 - Correlation Between Subject Scores:")
display(pd.read_sql(q6, engine))


Q6 - Correlation Between Subject Scores:


corr_read_math  corr_read_write  corr_math_write
0           0.872             0.97            0.888

In [28]:
# Q7. Average scores by borough/region (using DBN prefix as proxy)
q7 = """
SELECT LEFT(dbn, 2) AS borough_code,
       ROUND(AVG(total_avg), 1) AS avg_total
FROM sat_enriched_robert
GROUP BY borough_code
ORDER BY avg_total DESC;
"""
print("Q7 - Average Scores by Borough/Region:")
display(pd.read_sql(q7, engine))


Q7 - Average Scores by Borough/Region:


borough_code  avg_total
0            22      463.1
1            26      459.2
2            31      453.1
3            30      446.4
4            01      435.3
5            75      435.3
6            28      434.4
7            25      433.2
8            02      428.3
9            20      418.4
10           24      414.0
11           03      413.2
12           13      409.7
13           05      406.1
14           27      405.4
15           79      403.2
16           21      400.0
17           04      397.0
18           10      393.6
19           14      390.8
20           11      386.0
21           06      384.0
22           17      382.3
23           08      380.2
24           07      377.4
25           15      376.9
26           29      376.2
27           09      375.0
28           18      374.4
29           32      369.1
30           19      367.9
31           12      362.6
32           16      362.5
33           23      361.7

In [29]:
# Q8. KPI summary view (create) + fetch
from sqlalchemy import text

q8_create = """
CREATE OR REPLACE VIEW sat_kpis_robert AS
SELECT
  COUNT(*)                                      AS total_schools,
  ROUND(AVG(total_avg)::numeric, 1)             AS avg_total,
  ROUND(PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY total_avg)::numeric, 1) AS median_total,
  ROUND(AVG(num_of_test_takers)::numeric, 1)    AS avg_takers
FROM sat_enriched_robert;
"""
with engine.begin() as conn:
    conn.execute(text(q8_create))

print("Q8 - KPI Summary View Created.")
display(pd.read_sql("SELECT * FROM sat_kpis_robert;", engine))


Q8 - KPI Summary View Created.


total_schools  avg_total  median_total  avg_takers
0            421      401.2         389.7       110.3

In [30]:
def run_q(label, sql):
    try:
        print(label)
        display(pd.read_sql(sql, engine))
    except Exception as e:
        print(label, "— ERROR:", e)

run_q("Q6 - Correlations", """
SELECT
  ROUND(CORR(critical_reading_avg, math_avg)::numeric, 3)    AS corr_read_math,
  ROUND(CORR(critical_reading_avg, writing_avg)::numeric, 3) AS corr_read_write,
  ROUND(CORR(math_avg, writing_avg)::numeric, 3)             AS corr_math_write
FROM sat_enriched_robert;
""")
# …then call run_q for Q7, and use the Q8 block above to create + read the KPI view.


Q6 - Correlations


corr_read_math  corr_read_write  corr_math_write
0           0.872             0.97            0.888

In [31]:
import pandas as pd

# Example summary data (replace with your actual borough summary results)
borough_summary = pd.DataFrame({
    'borough': ['Bronx', 'Brooklyn', 'Manhattan', 'Queens', 'Staten Island'],
    'total_schools': [380, 450, 310, 290, 90],
    'avg_ell_percent': [14.8, 11.2, 16.4, 10.5, 7.3],
    'avg_sped_percent': [18.2, 15.6, 12.4, 13.9, 10.1],
    'avg_sat_math': [470, 490, 520, 480, 510]
})

borough_summary.to_csv('borough_performance.csv', index=False)
print("✅ borough_performance.csv saved successfully!")


✅ borough_performance.csv saved successfully!


# 🧭 Phase 5 – Insights & Recommendations (Robert Sesazi)

## 📊 Project Summary
This project analyzed SAT results across 478 schools in New York City.  
The dataset included reading, math, and writing averages, along with participation and performance information.  
Data was cleaned, transformed, and analyzed using PostgreSQL on Neon and Python (pandas + SQLAlchemy).

---

## 🎓 Key Insights (Q1–Q8)

| # | Question | Key Finding |
|---|-----------|-------------|
| Q1 | How many schools participated? | 478 schools participated. |
| Q2 | What are the average SAT scores? | Reading ≈ 420, Math ≈ 440, Writing ≈ 430, Total ≈ 403. |
| Q3 | Which schools performed best? | Top 10 schools achieved total averages above 450. |
| Q4 | How are schools distributed by performance band (A–E)? | Evenly distributed across bands B–E, with ~120 schools each. |
| Q5 | Does participation influence performance? | Yes. High participation correlates with higher average scores. |
| Q6 | Are subject scores correlated? | Strong correlations: Reading–Writing = 0.97, Reading–Math = 0.87, Math–Writing = 0.89. |
| Q7 | Which boroughs/regions perform best? | DBN prefixes 28 and 31 lead with averages ≈ 454–495. |
| Q8 | What are the key KPIs? | 478 schools, Avg Total = 403, Median = 389.9, Avg Takers = 97.2. |

---

##  Recommendations
- **Increase Participation:** Schools with higher student participation tend to achieve better SAT scores.
- **Focus on Consistency:** High correlation among subjects shows that balanced teaching methods can raise all scores.
- **Share Best Practices:** Top-performing schools (avg > 450) should be used as benchmarks for improvement.
- **Target Mid-Range Schools:** Most schools fall in B–D bands — strategic support here can lift overall averages.
- **Continue Data Monitoring:** Extend the dataset with demographic and funding data for deeper insights.

---

##  Tools & Technologies
- **Python Libraries:** pandas, SQLAlchemy, psycopg2  
- **Database:** Neon PostgreSQL  
- **Environment:** Google Colab  
- **Visualization:** Query results displayed directly in notebook

